libsel4vm/src/arch/arm/vgic/vgic.c

```diff
@@ -450,99 +445,83 @@ static inline int is_active(struct gic_dist_map *gic_dist, int irq, int vcpu_id)
     return is_spi_active(gic_dist, irq, vcpu_id);
 }
 
-static inline int vgic_add_overflow_cpu(struct lr_of *lr_overflow, struct virq_handle *irq)
+static inline int vgic_irq_enqueue(vgic_t *vgic, vm_vcpu_t *vcpu, struct virq_handle *irq)
 {
-    /* Add to overflow list */
-    int idx = lr_overflow->tail;
-    if (unlikely(lr_overflow->full)) {
-        ZF_LOGF("too many overflow irqs");
+    struct irq_queue *q = &vgic->irq_queue[vcpu->vcpu_id];
+
+    if (unlikely(IRQ_QUEUE_NEXT(q->tail) == q->head)) {
         return -1;
     }
 
-    lr_overflow->irqs[idx] = *irq;
-    lr_overflow->full = (lr_overflow->head == LR_OF_NEXT(lr_overflow->tail));
-    if (!lr_overflow->full) {
-        lr_overflow->tail = LR_OF_NEXT(idx);
-    }
+    q->irqs[q->tail] = irq;
+    q->tail = IRQ_QUEUE_NEXT(q->tail);
+
     return 0;
 }
 
-static inline int vgic_add_overflow(vgic_t *vgic, struct virq_handle *irq, vm_vcpu_t *vcpu)
+static inline struct virq_handle *vgic_irq_dequeue(vgic_t *vgic, vm_vcpu_t *vcpu)
 {
-    return vgic_add_overflow_cpu(&vgic->lr_overflow[vcpu->vcpu_id], irq);
-}
+    struct irq_queue *q = &vgic->irq_queue[vcpu->vcpu_id];
+    struct virq_handle *virq = NULL;
 
-static inline void vgic_handle_overflow_cpu(vgic_t *vgic, struct lr_of *lr_overflow, vm_vcpu_t *vcpu)
-{
-    /* copy tail, as vgic_vcpu_inject_irq can mutate it, and we do
-     * not want to process any new overflow irqs */
-    size_t tail = lr_overflow->tail;
-    for (size_t i = lr_overflow->head; i != tail; i = LR_OF_NEXT(i)) {
-        if (vgic_vcpu_inject_irq(vgic_dist, vcpu, &lr_overflow->irqs[i]) == 0) {
-            lr_overflow->head = LR_OF_NEXT(i);
-            lr_overflow->full = (lr_overflow->head == LR_OF_NEXT(lr_overflow->tail));
-        } else {
-            break;
-        }
+    if (q->head != q->tail) {
+        virq = q->irqs[q->head];
+        q->head = IRQ_QUEUE_NEXT(q->head);
     }
-}
 
-static inline void vgic_handle_overflow(vgic_t *vgic, vm_vcpu_t *vcpu)
-{
-    vgic_handle_overflow_cpu(vgic, &vgic->lr_overflow[vcpu->vcpu_id], vcpu);
+    return virq;
 }
 
-static int vgic_vcpu_inject_irq(struct vgic_dist_device *d, vm_vcpu_t *inject_vcpu, struct virq_handle *irq)
+static int vgic_find_empty_list_reg(vgic_t *vgic, vm_vcpu_t *vcpu)
 {
-    vgic_t *vgic;
-    int err;
     int i;
 
-    vgic = vgic_device_get_vgic(d);
-
-    seL4_CPtr vcpu;
-    vcpu = inject_vcpu->vcpu.cptr;
-    for (i = 0; i < MAX_LR_OVERFLOW; i++) {
-        if (vgic->irq[inject_vcpu->vcpu_id][i] == NULL) {
-            break;
+    for (i = 0; i < NUM_LIST_REGS; i++) {
```

Comment:
See we only need the loop variable inthe body now, so let's use c99 features:
```suggestion
    for (int i = 0; i < NUM_LIST_REGS; i++) {
```

Reply:
Did this on other places too.